File: server/app/api/v1/openrouter_proxy/openrouter_proxy.py
"""OpenRouter代理 - 第三方AI模型路由代理"""


import httpx
from fastapi import APIRouter, Body
from loguru import logger

from app.schemas.common import error, success
from app.utils.ai_keys import openrouter_key

router = APIRouter()

OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"


@router.post("/chat", operation_id="openrouter_chat", summary="OpenRouter对话")
async def chat(
    messages: list = Body(..., embed=True),
    model: str = Body("openai/gpt-4o-mini", embed=True),
    temperature: float = Body(0.7, embed=True),
    max_tokens: int = Body(2048, embed=True),
    api_key: str | None = None,
):
    async with httpx.AsyncClient(timeout=60) as client:
        try:
            headers = {"Content-Type": "application/json"}
            if openrouter_key(api_key):
                headers["Authorization"] = f"Bearer {api_key}"
            else:
                headers["HTTP-Referer"] = "https://ihui-ai.com"
                headers["X-Title"] = "ihui-ai-edu"
            r = await client.post(
                f"{OPENROUTER_BASE_URL}/chat/completions",
                json={
                    "model": model,
                    "messages": messages,
                    "temperature": temperature,
                    "max_tokens": max_tokens,
                },
                headers=headers,
            )
            return success(r.json())
        except Exception as e:
            logger.error(f"openrouter chat error: {e}")
            return error(str(e))


@router.post("/completion", operation_id="openrouter_completion", summary="OpenRouter文本补全")
async def completion(
    prompt: str = Body(..., embed=True),
    model: str = Body("openai/gpt-3.5-turbo-instruct", embed=True),
    max_tokens: int = Body(1024, embed=True),
    api_key: str | None = None,
):
    async with httpx.AsyncClient(timeout=60) as client:
        try:
            headers = {"Content-Type": "application/json"}
            if openrouter_key(api_key):
                headers["Authorization"] = f"Bearer {api_key}"
            r = await client.post(
                f"{OPENROUTER_BASE_URL}/completions",
                json={"model": model, "prompt": prompt, "max_tokens": max_tokens},
                headers=headers,
            )
            return success(r.json())
        except Exception as e:
            logger.error(f"openrouter completion error: {e}")
            return error(str(e))


@router.get("/models", operation_id="openrouter_models", summary="可用模型列表")
async def models():
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            r = await client.get(f"{OPENROUTER_BASE_URL}/models")
            return success(r.json())
    except Exception:
        return success(
            {
                "data": [
                    {"id": "openai/gpt-4o", "name": "GPT-4o"},
                    {"id": "openai/gpt-4o-mini", "name": "GPT-4o mini"},
                    {"id": "openai/gpt-3.5-turbo", "name": "GPT-3.5 Turbo"},
                    {"id": "anthropic/claude-3.5-sonnet", "name": "Claude 3.5 Sonnet"},
                    {"id": "google/gemini-pro-1.5", "name": "Gemini Pro 1.5"},
                ]
            }
        )


@router.post("/embeddings", operation_id="openrouter_embeddings", summary="OpenRouter Embeddings")
async def embeddings(
    input_text: str = Body(..., embed=True),
    model: str = Body("openai/text-embedding-3-small", embed=True),
    api_key: str | None = None,
):
    async with httpx.AsyncClient(timeout=60) as client:
        try:
            headers = {"Content-Type": "application/json"}
            if openrouter_key(api_key):
                headers["Authorization"] = f"Bearer {api_key}"
            r = await client.post(
                f"{OPENROUTER_BASE_URL}/embeddings",
                json={"model": model, "input": input_text},
                headers=headers,
            )
            return success(r.json())
        except Exception as e:
            logger.error(f"openrouter embeddings error: {e}")
            return error(str(e))


@router.get("/credits", summary="账户额度")
async def credits(api_key: str = ""):
    async with httpx.AsyncClient(timeout=30) as client:
        try:
            headers = {}
            if openrouter_key(api_key):
                headers["Authorization"] = f"Bearer {api_key}"
            r = await client.get(f"{OPENROUTER_BASE_URL}/credits", headers=headers)
            return success(r.json())
        except Exception:
            return success({"data": {"total_credits": 0, "total_usage": 0}})

File: client/src/api/openrouter-proxy.ts
/**
 * OpenRouter 代理 API
 * 对接后端: app/api/v1/openrouter_proxy/openrouter_proxy.py
 * 路由前缀: /api/v1/openrouter-proxy
 *
 * 后端 chat/completion/embeddings 使用 Body(embed=True),
 * 即请求体为 JSON 对象, 字段名与后端一致 (snake_case)。
 */
import http from '@/utils/request'
import type { ApiResponse } from '@/types'

/** OpenRouter 对话参数 */
export interface OpenRouterChatParams {
  messages: Array<{ role: string; content: string }>
  model: string
  temperature?: number
}

/** OpenRouter 文本补全参数 */
export interface OpenRouterCompletionParams {
  prompt: string
  model: string
  maxTokens?: number
}

/** OpenRouter Embedding 参数 */
export interface OpenRouterEmbeddingParams {
  inputText: string
  model: string
  apiKey?: string
}

function toDataResult(data: unknown, msg = 'success'): ApiResponse<unknown> {
  return {
    code: 0,
    message: msg,
    data,
    success: true,
    timestamp: Date.now(),
  } as unknown as ApiResponse<unknown>
}

// ===========================================================================
// OpenRouter 代理
// ===========================================================================

/** OpenRouter 对话 (Body 传参) */
export async function openrouterChat(params: OpenRouterChatParams): Promise<ApiResponse<unknown>> {
  const res = await http.post('/api/v1/openrouter-proxy/chat', {
    messages: params.messages,
    model: params.model,
    temperature: params.temperature,
  })
  const body = (res as any).data || {}
  return toDataResult(body.data, body.msg)
}

/** OpenRouter 文本补全 (Body 传参) */
export async function openrouterCompletion(params: OpenRouterCompletionParams): Promise<ApiResponse<unknown>> {
  const res = await http.post('/api/v1/openrouter-proxy/completion', {
    prompt: params.prompt,
    model: params.model,
    max_tokens: params.maxTokens,
  })
  const body = (res as any).data || {}
  return toDataResult(body.data, body.msg)
}

/** 可用模型列表 */
export async function openrouterModels(): Promise<ApiResponse<unknown>> {
  const res = await http.get('/api/v1/openrouter-proxy/models')
  const body = (res as any).data || {}
  return toDataResult(body.data, body.msg)
}

/** OpenRouter Embedding (Body 传参, apiKey 走 Query) */
export async function openrouterEmbeddings(params: OpenRouterEmbeddingParams): Promise<ApiResponse<unknown>> {
  const res = await http.post('/api/v1/openrouter-proxy/embeddings', {
    input_text: params.inputText,
    model: params.model,
  }, {
    params: { api_key: params.apiKey || undefined },
  })
  const body = (res as any).data || {}
  return toDataResult(body.data, body.msg)
}

/** 账户额度 (apiKey 走 Query) */
export async function openrouterCredits(params: {
  apiKey: string
}): Promise<ApiResponse<unknown>> {
  const res = await http.get('/api/v1/openrouter-proxy/credits', {
    params: { api_key: params.apiKey },
  })
  const body = (res as any).data || {}
  return toDataResult(body.data, body.msg)
}

export const openrouterProxyApi = {
  openrouterChat,
  openrouterCompletion,
  openrouterModels,
  openrouterEmbeddings,
  openrouterCredits,
}

export default openrouterProxyApi